polyfill/intersection-observer.js

```diff
@@ -267,9 +273,9 @@ IntersectionObserver.prototype._monitorIntersections = function() {
       addEvent(window, 'resize', this._checkForIntersections, true);
       addEvent(document, 'scroll', this._checkForIntersections, true);
 
-      if ('MutationObserver' in window) {
+      if (this.USE_MUTATION_OBSERVER && 'MutationObserver' in window) {
```

Comment:
As above, this shouldn't be publicly exposed if not in the spec, otherwise people using the polyfill might end up doing the wrong thing.

Reply:
This is the point of this flag, to be publicly exposed in order to deactivate mutation observer listeners which are very resource consuming and not needed in a lot of cases. (I would not say 99% but meh :p)

Reply:
You still risk authors becoming reliant on this non-standard behavior - which you definitely don't want in a polyfill. 

cc @triblondon for guidance.

Reply:
This polyfill is an implementation of the IntersectionObserver spec.
The use of the MutationObserver is an implementation detail and has nothing to do with the spec.
What @JSteunou has submitted is simply a way to configure this particular implementation which any user of the polyfill should be permitted to do.

I don't see how there is any risk in a user becoming reliant on this configuration property any more than they are already reliant on the polyfill to begin with.

Reply:
But it looks like it's modifying the interface definition on the global scope:

```JS
IntersectionObserver.prototype.USE_MUTATION_OBSERVER = true;
```

What am I missing? If my understanding is correct, then you've added something non-standard to the API.

Reply:
the configuration properties are orthogonal to the usage of the IntersectionObserver spec api.
the polyfill already provides 2 configuration properties and the [README](https://github.com/w3c/IntersectionObserver/tree/master/polyfill#configuring-the-polyfill) describes how to use them.

here's this new config property might be used:

my polyfills.ts
```
import 'intersection-observer/intersection-observer';
IntersectionObserver.prototype.USE_MUTATION_OBSERVER = false;
// if the browser natively supports IntersectionObserver, then the polyfill
// is a no-op and the property is moot
```

my app.ts
```
const options = {
  threshold: 0.25
};
const onIntersection = function(entries: IntersectionObserverEntry[], observer: IntersectionObserver): void {
  // do stuff with intersection entries
};
const observer = new IntersectionObserver(onIntersect, options);
// do stuff with observer
```

Reply:
@marcoscaceres if you're against exposing configuration flag on `IntersectionObserver` prototype then you should remove `THROTTLE_TIMEOUT` and `POLL_INTERVAL` for consistency.

Reply:
Yeah. Just thinking out loud here, but maybe put `POLL_INTERVAL` and friends into the container scope or as an `configOptions` on the bootstrap:
```js 
(function(window, document, configOptions) {
   const { THROTTLE_TIMEOUT,  POLL_INTERVAL, USE_MUTATION_OBSERVER } = configOptions; 
   // other stuff
}(window, document, {
  THROTTLE_TIMEOUT: false, 
  POLL_INTERVAL: false,  
  USE_MUTATION_OBSERVER: false,
}));
 ```

Reply:
How do you expect users to configure use of this polyfill this way? oO

Reply:
They could change the values manually, which is admittedly not great.  Another option would be to set it inline:

```HTML
<script>
var interOps = {
  THROTTLE_TIMEOUT: false, 
  POLL_INTERVAL: false,  
  USE_MUTATION_OBSERVER: true,
}
</script>
<script src="https://the.poly.fill/intersection.js"></script> 
```

And then the polyfill would have:

```JS
{
const defaultOps = {
  THROTTLE_TIMEOUT: false, 
  POLL_INTERVAL: false,  
  USE_MUTATION_OBSERVER: false,
};
(function(window, document, configOptions) {
  const { THROTTLE_TIMEOUT,  POLL_INTERVAL, USE_MUTATION_OBSERVER } = {
     // blend together 
     ... defaultOption, ...configOptions
  }; 
   // other stuff
}(window, document,  interOps ||  defaultOps));
```

Or some such. There are probably lots of ways of doing it - that's just one off the top of my head. 

Another might be using the URL of the script:

```HTML
<script src="https://the.poly.fill/intersection.js?THROTTLE_TIMEOUT=true"></script> 
```

Reply:
This is not as simple from a developer and user of this polyfill point of view. I'm speaking for my case but there are a lot of different usage & context. I'm using polyfill service and build with webpack, I do not add global variable neither query parameters on my build upon my app's views needs and usage, and I'm pretty sure everyone would have an example in which your changes would make them miserable.

Why one would make it difficult for everyone rather than keep it simple? I think the point of a polyfill is to help spread the use of new feature. If not, it would be never use until a decade.

In French we have a say I like a lot that can be like this in English 

> perfectionism can be counter-productive 

Meanwhile, why just not validate this PR, which solve an issue for users and you would make another one with your changes about those flag, moving this debate into the right place without blocking this one.

Reply:
Refactoring the configuration properties seems outside of the scope of this PR and is derailing.